stretch_demos/nodes/aruco_head_scan_action.py

```diff
@@ -0,0 +1,205 @@
+        self.goal = goal
+        self.aruco_id = self.goal.aruco_id
+        self.tilt_angle = self.goal.tilt_angle
+        self.fill_in_blindspot_with_second_scan = self.goal.fill_in_blindspot_with_second_scan
+        self.fast_scan = self.goal.fast_scan
+        self.publish_to_map = self.goal.publish_to_map
+        self.scan_and_detect()
+
+    def scan_and_detect(self):
+        self.aruco_tf = None
+        self.predock_tf = None
+        pan_angle = -3.69
+        
+        markers = []
+        scan_point = JointTrajectoryPoint()
+        scan_point.time_from_start = rospy.Duration(3.0)
+        scan_point.positions = [self.tilt_angle, pan_angle]
+
+        trajectory_goal = FollowJointTrajectoryGoal()
+        trajectory_goal.trajectory.joint_names = ['joint_head_tilt', 'joint_head_pan']
+        trajectory_goal.trajectory.points = [scan_point]
+        trajectory_goal.trajectory.header.stamp = rospy.Time.now()
+        trajectory_goal.trajectory.header.frame_id = 'base_link'
+
+        self.trajectory_client.send_goal(trajectory_goal)
+        self.trajectory_client.wait_for_result()
+
+        self.feedback.pan_angle = pan_angle
+        self.server.publish_feedback(self.feedback)
+
+        for pan_angle in np.arange(-3.69, 1.66, 0.7):
+            time.sleep(3.0)
+            for i in range(20):
+                if self.markers:
+                    markers = self.markers
+                    break
+
+            rospy.loginfo("Markers found: {}".format(markers))
+            
+            if markers != []:
+                for marker in markers:
+                    if marker.id == self.aruco_id:
+                        self.aruco_found = True
+                        self.aruco_name = marker.text
+                        if self.publish_to_map:
+                            trans = self.tf2_buffer.lookup_transform('map', self.aruco_name, rospy.Time())
+                            self.aruco_tf = self.broadcast_tf(trans.transform, self.aruco_name, 'map')
+                            rospy.loginfo("{} pose published to tf".format(self.aruco_name))
+                            if self.aruco_name == 'docking_station':
+                                # Transform the docking station frame such that x-axis points out of the aruco plane and 0.5 m in the front of the dock
+                                # This facilitates passing the goal pose as this predock frame so that the robot can back up into the dock
+                                saved_pose = Transform()
+                                saved_pose.translation.x = 0.0
+                                saved_pose.translation.y = -0.45
+                                saved_pose.translation.z = 0.47
+                                saved_pose.rotation.x = -0.382
+                                saved_pose.rotation.y = -0.352
+                                saved_pose.rotation.z = -0.604
+                                saved_pose.rotation.w = 0.604
```

Comment:
Will it be appropriate/possible to make these hardcoded offsets for translations/rotations to project the predock pose at a distance from the dock pose as a variable?

Reply:
Since this is a saved pose, it will have to be supplied like this or in a similar manner. I obtained this pose by detecting the docking station marker, setting the robot in predock pose manually, and then doing a TF transform between the base_link and the docking_station.

Do you have a better idea?